src/Changelog/ReleaseChangelogViaKeepAChangelog.php

```diff
@@ -0,0 +1,141 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Laminas\AutomaticReleases\Changelog;
+
+use Laminas\AutomaticReleases\Git\CommitFile;
+use Laminas\AutomaticReleases\Git\Push;
+use Laminas\AutomaticReleases\Git\Value\BranchName;
+use Laminas\AutomaticReleases\Git\Value\SemVerVersion;
+use Phly\KeepAChangelog\Common\DiscoverChangelogEntryListener;
+use Phly\KeepAChangelog\Config;
+use Phly\KeepAChangelog\Version\ReadyLatestChangelogEvent;
+use Phly\KeepAChangelog\Version\SetDateForChangelogReleaseListener;
+use Symfony\Component\Console\Output\NullOutput;
+
+use function date;
+use function file_exists;
+use function sprintf;
+
+final class ReleaseChangelogViaKeepAChangelog implements ReleaseChangelog
+{
+    private CommitFile $commitFile;
+    private Push $push;
+
+    public function __construct(
+        CommitFile $commitFile,
+        Push $push
+    ) {
+        $this->commitFile = $commitFile;
+        $this->push       = $push;
+    }
+
+    /**
+     * @psalm-param non-empty-string $repositoryDirectory
+     */
+    public function __invoke(
+        string $repositoryDirectory,
+        SemVerVersion $version,
+        BranchName $sourceBranch
+    ): void {
+        $changelogFile = $repositoryDirectory . '/CHANGELOG.md';
+        if (! file_exists($changelogFile)) {
+            // No changelog
+            return;
+        }
+
+        $versionString = $version->fullReleaseName();
+
+        if (! $this->updateChangelog($changelogFile, $versionString)) {
+            // Failure to update; nothing to commit
+            return;
+        }
+
+        ($this->commitFile)(
+            $repositoryDirectory,
+            $sourceBranch,
+            'CHANGELOG.md',
+            sprintf('%s readiness', $versionString)
+        );
+
+        ($this->push)($repositoryDirectory, $sourceBranch->name());
+    }
+
+    private function updateChangelog(string $changelogFile, string $versionString): bool
+    {
+        $event = $this->createReadyLatestChangelogEvent($changelogFile, $versionString);
+
+        (new DiscoverChangelogEntryListener())($event);
+
+        if ($event->failed()) {
+            return false;
+        }
+
+        (new SetDateForChangelogReleaseListener())($event);
+
+        return ! $event->failed();
+    }
+
+    private function createReadyLatestChangelogEvent(
+        string $changelogFile,
+        string $versionString
+    ): ReadyLatestChangelogEvent {
+        /**
+         * Hard-coded extension to allow setting version to known value
+         *
+         * @psalm-suppress PropertyNotSetInConstructor
+         */
+        $event = new class ($versionString) extends ReadyLatestChangelogEvent {
```

Comment:
I used anonymous classes in this implementation as they are one-offs, and only set the values necessary for the specific event listeners I'm manually triggering. If you want me to extract these, let me know.

Reply:
I think this is an OK trade-off: we can't go and change the whole upstream library for this :)